src/refiners/foundationals/swin/mvanet/mclm.py

```diff
@@ -0,0 +1,211 @@
+        pos_y = y_embed / self.dim_t
+
+        pos_x = torch.stack((pos_x[:, :, :, 0::2].sin(), pos_x[:, :, :, 1::2].cos()), dim=4).flatten(3)
+        pos_y = torch.stack((pos_y[:, :, :, 0::2].sin(), pos_y[:, :, :, 1::2].cos()), dim=4).flatten(3)
+        return torch.cat((pos_y, pos_x), dim=3).permute(1, 2, 0, 3).flatten(0, 1)
+
+
+class MultiPoolPos(fl.Module):
+    def __init__(self, pool_ratios: list[int], positional_embedding: PositionEmbeddingSine):
+        super().__init__()
+        self.pool_ratios = pool_ratios
+        self.positional_embedding = positional_embedding
+
+    def forward(self, *args: int) -> Tensor:
+        h, w = args
+        return torch.cat([self.positional_embedding(h // ratio, w // ratio) for ratio in self.pool_ratios])
+
+
+class Repeat(fl.Module):
+    def __init__(self, dim: int = 0):
+        self.dim = dim
+        super().__init__()
+
+    def forward(self, x: Tensor, n: int) -> Tensor:
+        return torch.repeat_interleave(x, n, dim=self.dim)
+
+
+class _MHA_Arg(fl.Sum):
+    def __init__(self, offset: int):
+        self.offset = offset
+        super().__init__(
+            fl.GetArg(offset),  # value
+            fl.Chain(
+                fl.Parallel(
+                    fl.GetArg(self.offset + 1),  # position embedding
+                    fl.Lambda(self._batch_size),
+                ),
+                Repeat(1),
+            ),
+        )
+
+    def _batch_size(self, *args: Tensor) -> int:
+        return args[self.offset].size(1)
+
+
+class GlobalAttention(fl.Chain):
+    # Input must be a 4-tuple: (global, global pos. emb, pools, pools pos. emb.)
+    def __init__(
+        self,
+        emb_dim: int,
+        num_heads: int = 1,
+        device: Device | None = None,
+    ):
+        super().__init__(
+            fl.Sum(
+                fl.GetArg(0),  # global
+                fl.Chain(
+                    fl.Parallel(
+                        _MHA_Arg(0),  # Q: global + pos. emb
+                        _MHA_Arg(2),  # K: pools + pos. emb
+                        fl.GetArg(2),  # V: pools
+                    ),
+                    MultiheadAttention(emb_dim, num_heads, device=device),
+                ),
+            ),
+        )
+
+
+class MCLM(fl.Chain):
+    """Multi-View Complementary Localization Module
+    Inputs:
+        tensor: (b, 5, e, h, h)
+    Outputs:
+        tensor: (b, 5, e, h, h)
+    """
+
+    def __init__(
+        self,
+        emb_dim: int,
+        num_heads: int = 1,
+        pool_ratios: list[int] | None = None,
+        device: Device | None = None,
+    ):
+        if pool_ratios is None:
+            pool_ratios = [2, 8, 16]
+
+        positional_embedding = PositionEmbeddingSine(num_pos_feats=emb_dim // 2, device=device)
+
+        # LayerNorms in MCLM share their weights.
+
+        ln1 = fl.LayerNorm(emb_dim, device=device)
+        ln2 = fl.LayerNorm(emb_dim, device=device)
+
+        def proxy(m: fl.Module) -> fl.Module:
```

Comment:
I was tempted to add another comment (or expand the above one re: layer norms shared weights), just to make it clear which role `proxy` has there.